Refactoring help component (and fragmenting it too)

Rafi993 opened the pull request:
## What type of PR is this?

- [x] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
Refactoring Help component (and fragmenting it too) for improving readability. Help components had function that were returning component like
```
  renderBasicEditorHelp = () =>
   <div>
    ...
   </div>

```
I have moved them into their own files with logic to decide if those components should render inside them.

## Related Tickets & Documents
Closes https://github.com/forem/forem/issues/10949

## QA Instructions, Screenshots, Recordings

## Added tests?

Since it is a refactor the tests for this code is already there

- [ ] yes
- [x] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [ ] docs.forem.com
- [ ] readme
- [x] no documentation needed

Rafi993 commented:
I think the code climate error might be a false positive.

nickytonline commented:
@Rafi993, it looks like when you merged master into your branch something went wrong. There are files not related to your changes that are included in your PR.

Rafi993 commented:
> @Rafi993, it looks like when you merged master into your branch something went wrong. There are files not related to your changes that are included in your PR.

Thank you @nickytonline let me check again.

Rafi993 commented:
> > @Rafi993, it looks like when you merged master into your branch something went wrong. There are files not related to your changes that are included in your PR.
> 
> @nickytonline I just checked it again I think they are files that I changed. I fragmented Help component into separate components. Help components had function that were returning component like
> 
> ```
>   renderBasicEditorHelp = () =>
>    <div>
>     ...
>    </div>
> ```
> 
> I moved them into files of their own as components.

I think I should clear PR descriptions.

nickytonline commented on app/javascript/article-form/components/Help/index.jsx:
```diff
@@ -0,0 +1,122 @@
+
+  toggleModal = (varShowing) => (e) => {
+    e.preventDefault();
+    this.setState((prevState) => ({
+      ...this.setCommonProps({
+        [varShowing]: !prevState[varShowing],
+      }),
+    }));
```
refactor (blocking): `setCommonProps` is only ever called to spread the intial state of some state properties in `toggleModal`. Why not make it part of the initial state and then you can use `prevState` to get the value you need to set the key `[varShiowing]`.

```suggestion
    this.setState((prevState) => ({
      [varShowing]: !prevState[varShowing],
    }));
```

nickytonline commented on app/javascript/article-form/components/Help/Format.jsx:
```diff
@@ -0,0 +1,135 @@
+import PropTypes from 'prop-types';
+
+const Format = ({ show, toggleModal }) => {
+  if (show) {
```
refactor (blocking): For a few of the components that do `if (show) { ...` to render, move the responsibility of showing or hiding the component to the parent component.

nickytonline commented on app/javascript/article-form/components/Help/Format.jsx:
```diff
@@ -0,0 +1,135 @@
+import { h } from 'preact';
+import PropTypes from 'prop-types';
+
+const Format = ({ show, toggleModal }) => {
```
refactor (non-blocking): Consider a better name for the component. I'm not sure what the `<Format />` component is by looking at it. Looking at the JSX within, I can deduce it's for displaying help. Maybe `<EditorHelp />` or `<EditorFormattingHelp />`

nickytonline reviewed (changes requested):
Still need to review this some more @Rafi993, but I do have a few change requests. Thanks for being patient. I've been pretty busy during Hacktoberfest.

Rafi993 commented:
> Still need to review this some more @Rafi993, but I do have a few change requests. Thanks for being patient. I've been pretty busy during Hacktoberfest.

Thanks for your time @nickytonline

msarit commented:
@Rafi993 Hello! Thanks for this PR - could you please resolve the conflicts? Thanks! 🙏🏾

Rafi993 commented:
> @Rafi993 Hello! Thanks for this PR - could you please resolve the conflicts? Thanks! 🙏🏾

I have fixed merge conflicts. There seems to be some unrelated rspec tests failing.

nickytonline commented:
Things still behave as expected. Just going to give this a review.

nickytonline commented on app/javascript/article-form/components/Help/index.jsx:
```diff
@@ -0,0 +1,109 @@
+import { h, Component } from 'preact';
```
Refactor (blocking): Since this has been pulled out for the refactor, might as well convert it to use hooks. Here's a rough working implementation. Feel free to tweak it.

Also, since we've refactored this, it would be good to add JSDoc comments for the component.

```jsx
import { h } from 'preact';
import { useState } from 'preact/hooks';
import PropTypes from 'prop-types';
import ArticleFormTitle from './ArticleFormTitle';
import TagInput from './TagInput';
import BasicEditor from './BasicEditor';
import EditorFormattingHelp from './EditorFormattingHelp';
import { Modal } from '@crayons';

function renderModal(onClose, title, helpHtml) {
  return (
    <Modal onClose={onClose} title={title}>
      <div
        // eslint-disable-next-line react/no-danger
        dangerouslySetInnerHTML={{ __html: helpHtml }}
      />
    </Modal>
  );
}

export function Help({ previewShowing, helpFor, helpPosition, version }) {
  const [state, setState] = useState({
    liquidHelpHTML:
      document.getElementById('editor-liquid-help') &&
      document.getElementById('editor-liquid-help').innerHTML,
    markdownHelpHTML:
      document.getElementById('editor-markdown-help') &&
      document.getElementById('editor-markdown-help').innerHTML,
    frontmatterHelpHTML:
      document.getElementById('editor-frontmatter-help') &&
      document.getElementById('editor-frontmatter-help').innerHTML,
    liquidShowing: false,
    markdownShowing: false,
    frontmatterShowing: false,
  });

  function toggleModal(varShowing) {
    return function (e) {
      e.preventDefault();

      setState((prevState) => ({
        ...prevState,
        [varShowing]: !prevState[varShowing],
      }));
    };
  }

  const {
    liquidHelpHTML,
    markdownHelpHTML,
    frontmatterHelpHTML,
    liquidShowing,
    markdownShowing,
    frontmatterShowing,
  } = state;

  return (
    <div className="crayons-article-form__aside">
      {!previewShowing && (
        <div
          data-testid="article-form__help-section"
          className="sticky"
          style={{ top: version === 'v1' ? '56px' : helpPosition }}
        >
          {helpFor === 'article-form-title' && <ArticleFormTitle />}
          {helpFor === 'tag-input' && <TagInput />}

          {version === 'v1' && <BasicEditor toggleModal={toggleModal} />}

          {(helpFor === 'article_body_markdown' || version === 'v1') && (
            <EditorFormattingHelp toggleModal={toggleModal} />
          )}
        </div>
      )}

      {liquidShowing &&
        renderModal(
          toggleModal('liquidShowing'),
          '🌊 Liquid Tags',
          liquidHelpHTML,
        )}

      {markdownShowing &&
        renderModal(
          toggleModal('markdownShowing'),
          '✍️ Markdown',
          markdownHelpHTML,
        )}

      {frontmatterShowing &&
        renderModal(
          toggleModal('frontmatterShowing'),
          'Jekyll Front Matter',
          frontmatterHelpHTML,
        )}
    </div>
  );
}

Help.propTypes = {
  previewShowing: PropTypes.bool.isRequired,
  helpFor: PropTypes.string.isRequired,
  helpPosition: PropTypes.string.isRequired,
  version: PropTypes.string.isRequired,
};

Help.displayName = 'Help';
```

Rafi993 replied:
I have converted it to functional component now

nickytonline reviewed (changes requested):
Thanks for the PR @Rafi993! This looks great. Just some change requests.

msarit commented:
Hey @Rafi993! 👋🏾 
Just drawing your attention to this PR; looks like Nick left some feedback for you.
Let us know if you have questions or need help.

aitchiss commented:
Hi @Rafi993 - just checking in to see if you're still working on this one? Let me know if you need any assistance!

Rafi993 commented:
> Hi @Rafi993 - just checking in to see if you're still working on this one? Let me know if you need any assistance!

I'll fix conflicts in the PR today and then work on the changes tomorrow.

Rafi993 commented:
> > Hi @Rafi993 - just checking in to see if you're still working on this one? Let me know if you need any assistance!
> 
> I'll fix conflicts in the PR today and then work on the changes tomorrow.

Sorry I was not able to look at this on Monday I'll take a look at it tomorrow. I have fixed the merge conflicts now.

citizen428 commented:
Thanks, @Rafi993, your contributions are always much appreciated. Please feel free to tag us once you're done or if you need any additional help or feedback.

aitchiss reviewed (changes requested):
Hi @Rafi993 - I tried giving this a quick spin locally but it seems there's an issue with the modal toggling:

![Screenshot showing empty modal](https://user-images.githubusercontent.com/20773163/112129864-c086b480-8bbf-11eb-9f81-1fd9676d6c07.png)

I get an empty modal, and then I can't close it either with the button or the Escape key. Can you take a look? Let me know if you need any assistance!

Rafi993 commented:
> Hi @Rafi993 - I tried giving this a quick spin locally but it seems there's an issue with the modal toggling:
> 
> ![Screenshot showing empty modal](https://user-images.githubusercontent.com/20773163/112129864-c086b480-8bbf-11eb-9f81-1fd9676d6c07.png)
> 
> I get an empty modal, and then I can't close it either with the button or the Escape key. Can you take a look? Let me know if you need any assistance!

Let me look into that tommorow.

Rafi993 commented:
> Hi @Rafi993 - I tried giving this a quick spin locally but it seems there's an issue with the modal toggling:
> 
> ![Screenshot showing empty modal](https://user-images.githubusercontent.com/20773163/112129864-c086b480-8bbf-11eb-9f81-1fd9676d6c07.png)
> 
> I get an empty modal, and then I can't close it either with the button or the Escape key. Can you take a look? Let me know if you need any assistance!

Thank you for your patience @aitchiss I have fixed the issue now. This is how it renders now 

<img width="1680" alt="Screenshot 2021-03-29 at 8 33 16 AM" src="https://user-images.githubusercontent.com/8092120/112781980-8c9d0a80-9069-11eb-8644-e8bb7d075341.png">

aitchiss commented on app/javascript/article-form/components/Help/index.jsx:
```diff
@@ -0,0 +1,94 @@
+
+const renderModal = (onClose, title, selector) => {
+  const helpHtml =
+    document.getElementById(selector) &&
+    document.getElementById(selector).innerHTML;
```
suggestion (not-blocking) - I think you can use optional chaining to get the same effect here

```suggestion
    document.getElementById(selector)?.innerHTML;
```

aitchiss commented on app/javascript/article-form/components/Help/index.jsx:
```diff
@@ -0,0 +1,94 @@
+    frontmatterShowing: false,
+  });
+
+  const openModal = (varShowing) => {
```
```suggestion
  const openModal = (helpSection) => {
```

suggestion (not-blocking) - while we are refactoring these files I think it would be beneficial to improve some of the arg names here to better reflect what's being passed in (same with `closeModal`)

aitchiss commented on app/javascript/article-form/components/Help/index.jsx:
```diff
@@ -0,0 +1,94 @@
+};
+
+export const Help = ({ previewShowing, helpFor, helpPosition, version }) => {
+  const [state, setState] = useState({
```
suggestion (not-blocking) - consider giving this a more specific name than `state` - maybe something like `helpSectionVisibilityState` (maybe you can think of something better than this though!)

aitchiss reviewed (approved):
Thanks for the work on this one! I've left some suggestions that I think could clean this up even more, but I've tested locally and this is looking good 👍

nickytonline commented on app/javascript/article-form/components/Help/index.jsx:
```diff
@@ -0,0 +1,94 @@
+  );
+};
+
+export const Help = ({ previewShowing, helpFor, helpPosition, version }) => {
```
Documentation (blocking): As mentioned in https://github.com/forem/forem/pull/10895#discussion_r552984391, add some JS Doc comments since we are refactoring this to a function component and it is exported. Same for the other exported components in this PR.

nickytonline reviewed (changes requested):
@Rafi993, this looks good. I tested it locally and it's still behaving the same way. A small change request for documentation, but aside from that, this is good to go.

Thanks so much for this contribution. Looking forward to seeing it in production. 😎

citizen428 commented:
Hi @Rafi993, I just wanted to check in and see what the status of this PR is? Are you planning on adding the docs @nickytonline requested? Thanks as always for your contributions!

Rafi993 commented:
> Hi @Rafi993, I just wanted to check in and see what the status of this PR is? Are you planning on adding the docs @nickytonline requested? Thanks as always for your contributions!

Hi, I have made the suggested changes now.

citizen428 commented:
Thank you so much @Rafi993!

nickytonline reviewed (approved):
As always, thanks for tour contributions @Rafi993!
